lte/cloud/go/services/subscriberdb/storage/subscriber_state_test.go

```diff
@@ -101,4 +109,90 @@ func TestSubscriberStorage(t *testing.T) {
+		imsisGw2 := []string{"IMSI002020000000123", "IMSI002020000000456"}
+		subscriberStatesGw1 := CreateTestGatewaySubscriberState(imsisGw1...)
+		subscriberStatesGw2 := CreateTestGatewaySubscriberState(imsisGw2...)
+
+		err = s.SetAllSubscribersForGateway(nid1, gwid1, &subscriberStatesGw1)
+		assert.NoError(t, err)
+		err = s.SetAllSubscribersForGateway(nid1, gwid2, &subscriberStatesGw2)
+		assert.NoError(t, err)
+		err = s.SetAllSubscribersForGateway(nid2, gwid1, &subscriberStatesGw1)
+		assert.NoError(t, err)
+
+		// get all IMSIs for network
+
+		foundStates, err := s.GetSubscribersForIMSIs(nid1, nil)
+		assert.NoError(t, err)
+		maps.Copy(subscriberStatesGw1.Subscribers, subscriberStatesGw2.Subscribers)
+		assert.True(t, cmp.Equal(foundStates, subscriberStatesGw1.Subscribers))
+
+		// get selected IMSIs for network (from two gateways)
+
+		selectedIMSIs := []string{"IMSI001010000000123", "IMSI002020000000456"}
+		expectedStates := map[string]state.ArbitraryJSON{
+			"IMSI001010000000123": subscriberStatesGw1.Subscribers["IMSI001010000000123"],
+			"IMSI002020000000456": subscriberStatesGw2.Subscribers["IMSI002020000000456"],
+		}
+		foundStates, err = s.GetSubscribersForIMSIs(nid1, selectedIMSIs)
+		assert.NoError(t, err)
+		assert.True(t, cmp.Equal(foundStates, expectedStates))
+
+		// get latest entry for IMSI appearing in two gateways
+		// the order of the imsis in imsisGw2 is important. The target
+		// doubled imsi "IMSI001010000000123", has to be the second entry,
+		// since its two entries are distinguished by the value of
+		// "active_duration_sec", which wouldn't differ otherwise
+
+		imsisGw2 = []string{"IMSI002020000000456", "IMSI001010000000123"}
+		subscriberStatesGw2 = CreateTestGatewaySubscriberState(imsisGw2...)
+
+		active_duration_sec_old := subscriberStatesGw1.Subscribers["IMSI001010000000123"]["magma.ipv4"].([]interface{})[0].(map[string]interface{})["active_duration_sec"]
+		active_duration_sec_new := subscriberStatesGw2.Subscribers["IMSI001010000000123"]["magma.ipv4"].([]interface{})[0].(map[string]interface{})["active_duration_sec"]
+		assert.Equal(t, active_duration_sec_old, float64(0))
+		assert.Equal(t, active_duration_sec_new, float64(1))
+
+		err = s.(*subscriberStorage).setTestGatewaySubscriberStateWithIncreasingTimestamp(nid1, gwid2, &subscriberStatesGw2)
+		assert.NoError(t, err)
+		foundStates, err = s.GetSubscribersForIMSIs(nid1, selectedIMSIs)
+		assert.NoError(t, err)
+		assert.True(t, cmp.Equal(foundStates, subscriberStatesGw2.Subscribers))
+		active_duration_sec_found := foundStates["IMSI001010000000123"]["magma.ipv4"].([]interface{})[0].(map[string]interface{})["active_duration_sec"]
+		assert.Equal(t, active_duration_sec_found, float64(1))
+	})
+}
+
+func (ss *subscriberStorage) setTestGatewaySubscriberStateWithIncreasingTimestamp(networkID string, gatewayID string, subscriberStates *GatewaySubscriberState) error {
+	timeSec := clock.Now()
```

Comment:
Currently this method is only used to set the states in the final case of the test, the duplicate IMSIs. It uses system time (as `SetAllSubscribersForGateway`) and adds one second for every entry.
I.e. the first entry of `double_IMSI` is registered at system time (via `SetAllSubscribersForGateway`), the duplicate is registered at system time + 2 Seconds (via this method, `setTestGatewaySubscriberStateWithIncreasingTimestamp`).

More work, but may be less confusing could be to fix the time and set all states via this method? What do you think?

Reply:
Yes, I think it would be clearer to just give them explicit times and not rely on the system time + x seconds, and to create both states by the same method.